c10/core/TensorImpl.h

```diff
@@ -228,20 +229,62 @@ struct C10_API VariableVersion {
   c10::intrusive_ptr<VersionCounter> version_counter_;
 
  public:
+  // It's okay to return true even for inference tensor which
+  // doesn't have version counter enabled.
   bool unique() const {
-    return 1 == version_counter_.use_count();
+    return this->enabled() ? 1 == version_counter_.use_count() : true;
   }
   // NOTE: As of C++11 and 14, default-constructing a std::atomic variable
   // leaves it in a persistently undefined state. See
   // https://cplusplus.github.io/LWG/issue2334.
   VariableVersion(uint32_t version = 0)
       : version_counter_(c10::make_intrusive<VersionCounter>(version)) {}
+  explicit VariableVersion(bool enabled, uint32_t version = 0) {
```

Comment:
This is not a very safe constructor, because C++ is willing to convert freely between unsigned int and bool. https://godbolt.org/z/az5jGaofx Additionally, `versions` is ignored when `!enabled`, which makes an illegal state representable.

A better way of doing two variant constructors is to use the scoped enum trick. You can see an example of this in c10/core/DispatchKeySet.h